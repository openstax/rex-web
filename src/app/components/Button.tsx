import React from 'react';
import styled, { css } from 'styled-components/macro';
import { isDefined } from '../guards';
import theme, { ColorSet } from '../theme';
import { decoratedLinkStyle, linkColor, linkStyle } from './Typography';
import { textStyle } from './Typography/base';

const applyColor = (color: ColorSet) => `

  color: ${color.foreground};
  background-color: ${color.base};

  ${color.darker && css`
    :hover {
      background-color: ${color.darker};
    }
  `}
  ${color.darkest && css`
    :active {
      background-color: ${color.darkest};
    }
  `}
`;

type Variant = 'primary' | 'secondary' | 'transparent' | 'default';
type Size = 'large' | 'medium' | 'small';

type ComponentType = keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;

interface ButtonProps<T extends ComponentType | undefined> {
  disabled?: boolean;
  variant?: Variant;
  size?: Size;
  className?: string;
  component?: T extends undefined ? undefined :
    T extends ComponentType ? React.ReactComponentElement<T>:
    never;
}

// tslint:disable-next-line:variable-name
const ButtonHoc = React.forwardRef(<T extends ComponentType | undefined>(
  {variant, size, component, ...props}: ButtonProps<T>,
  ref: React.Ref<Omit<HTMLButtonElement | T, 'undefined'>>
) => {
  if (isDefined(component)) {
    return React.cloneElement(component, {...props, ref});
  }
  return <button ref={ref} {...props} />;
});

// tslint:disable-next-line:variable-name
const Button = styled(ButtonHoc)`
  display: flex;
  cursor: pointer;
  align-items: center;
  justify-content: center;
  border-radius: 0.2rem;
  text-decoration: none;
  white-space: nowrap;
  overflow: initial;
  ${(props) => props.size === 'large' && `
    font-size: 1.6rem;
    height: 5rem;
    padding: 0 3rem;
  `}
  ${(props) => (props.size === 'medium' || props.size === undefined) && `
    min-width: 12rem;
    font-size: 1.6rem;
    height: 4rem;
    padding: 0 3rem;
  `}
  ${(props) => props.size === 'small' && `
    font-size: 1.4rem;
    height: 3rem;
    padding: 0 2rem;
  `}

  ${(props) => props.variant === 'primary' && `
    ${applyColor(theme.color.primary.orange)}
    font-weight: bold;
    border: none;
  `}
  ${(props) => props.variant === 'secondary' && `
    ${applyColor(theme.color.secondary.lightGray)}
    font-weight: bold;
    border: none;
  `}
  ${(props) => props.variant === 'transparent' && `
    border: none;
    background-color: transparent;
    color: ${linkColor};
    font-weight: normal;
  `}
  ${(props) => (props.variant === 'default' || props.variant === undefined) && `
    ${applyColor({
      ...theme.color.neutral,
      foreground: theme.color.primary.gray.base,
    })}
    border: 1px solid #d5d5d5;
    font-weight: regular;
  `}
  ${(props) => props.disabled && `
    ${applyColor(theme.color.disabled)}
    cursor: not-allowed;
  `}
`;

// tslint:disable-next-line:variable-name
export const ButtonGroup = styled.div`
  display: grid;
  overflow: visible;
  ${(props: {expand?: boolean}) => props.expand === false && css`
    grid-auto-columns: min-content;
  `}
  grid-auto-flow: column;
  grid-gap: 1rem;
`;

// tslint:disable-next-line:variable-name
export const PlainButton = styled.button`
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  background: none;
`;

// tslint:disable-next-line:variable-name
export const ButtonLink = styled(PlainButton)`
  outline: none;
  ${textStyle}
  ${(props: {decorated: boolean}) => props.decorated ? decoratedLinkStyle : linkStyle}
`;

export default Button;
