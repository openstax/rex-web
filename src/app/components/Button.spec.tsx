import React from 'react';
import renderer from 'react-test-renderer';
import Button from './Button';
import { ButtonGroup, ButtonLink } from './Button';

describe('Button', () => {
  it('matches snapshot - link', () => {
    const component = renderer.create(<Button component={<a href='/' />}>this is a link</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('matches snapshot - small', () => {
    const component = renderer.create(<Button size='small' />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('matches snapshot - medium', () => {
    const component = renderer.create(<Button size='medium' />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('matches snapshot - large', () => {
    const component = renderer.create(<Button size='large' />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('matches snapshot - primary', () => {
    const component = renderer.create(<Button variant='primary' />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('matches snapshot - secondary', () => {
    const component = renderer.create(<Button variant='secondary' />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('matches snapshot - default', () => {
    const component = renderer.create(<Button variant='default' />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ButtonLink', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<ButtonLink />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot (decorated)', () => {
    const component = renderer.create(<ButtonLink decorated />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot (decorated and disabled)', () => {
    const component = renderer.create(<ButtonLink decorated disabled />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ButtonGroup', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<ButtonGroup />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot (expand)', () => {
    const component = renderer.create(<ButtonGroup expand />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot (not-expand)', () => {
    const component = renderer.create(<ButtonGroup expand={false} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
