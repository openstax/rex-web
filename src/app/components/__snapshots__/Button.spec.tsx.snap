// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot - default 1`] = `
<button
  className="Button-ayg7nk-0 liRzrl"
/>
`;

exports[`Button matches snapshot - large 1`] = `
<button
  className="Button-ayg7nk-0 dSJsyM"
/>
`;

exports[`Button matches snapshot - link 1`] = `
<a
  className="Button-ayg7nk-0 liRzrl"
  href="/"
>
  this is a link
</a>
`;

exports[`Button matches snapshot - medium 1`] = `
<button
  className="Button-ayg7nk-0 liRzrl"
/>
`;

exports[`Button matches snapshot - primary 1`] = `
<button
  className="Button-ayg7nk-0 fFJxez"
/>
`;

exports[`Button matches snapshot - secondary 1`] = `
<button
  className="Button-ayg7nk-0 dYeAnt"
/>
`;

exports[`Button matches snapshot - small 1`] = `
<button
  className="Button-ayg7nk-0 bHTkEr"
/>
`;

exports[`ButtonLink matches snapshot (decorated and disabled) 1`] = `
<button
  className="Button__PlainButton-ayg7nk-2 Button__ButtonLink-ayg7nk-3 Rtfaq"
  disabled={true}
/>
`;

exports[`ButtonLink matches snapshot (decorated) 1`] = `
<button
  className="Button__PlainButton-ayg7nk-2 Button__ButtonLink-ayg7nk-3 flLYdq"
/>
`;

exports[`ButtonLink matches snapshot 1`] = `
<button
  className="Button__PlainButton-ayg7nk-2 Button__ButtonLink-ayg7nk-3 cXLkqS"
/>
`;
